File: src/main/java/org/pyjjs/scheduler/core/api/impl/actors/system/SchedulingController.java
package org.pyjjs.scheduler.core.api.impl.actors.system;

import com.google.common.collect.Sets;
import org.pyjjs.scheduler.core.api.SchedulingListener;
import org.pyjjs.scheduler.core.api.impl.actors.common.behaviours.BehaviourBasedActor;
import org.pyjjs.scheduler.core.api.impl.actors.system.behaviours.GotNewChangesBehaviour;
import org.pyjjs.scheduler.core.api.impl.actors.system.behaviours.NotifyAboutChangesBehaviour;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class SchedulingController extends BehaviourBasedActor<SchedulingControllerState> {

    private static final Logger LOG = LoggerFactory.getLogger(SchedulingController.class);

    public SchedulingController(SchedulingListener listeners) {
        SchedulingControllerState state = getCopyOfActorState();
        state.setSchedulingListeners(Sets.newHashSet(listeners));
        updateActorState(state);
    }

    @Override
    protected SchedulingControllerState getInitialState() {
        return new SchedulingControllerState(getContext());
    }

    @Override
    protected void init() {
        fillBehaviours();
        LOG.info("Scheduling controller is initialized");
    }

    private void fillBehaviours() {
        addBehaviour(GotNewChangesBehaviour.class);
        addBehaviour(NotifyAboutChangesBehaviour.class);
    }
}

File: src/main/kotlin/org/pyjjs/scheduler/core/api/impl/actors/system/SchedulingController.kt
package org.pyjjs.scheduler.core.api.impl.actors.system

import org.pyjjs.scheduler.core.api.RegistryOfStrategies
import org.pyjjs.scheduler.core.api.SchedulingListener
import org.pyjjs.scheduler.core.api.impl.actors.common.behaviours.BehaviourBasedActor
import org.pyjjs.scheduler.core.api.impl.actors.system.behaviours.GotNewChangesBehaviour
import org.pyjjs.scheduler.core.api.impl.actors.system.behaviours.NotifyAboutChangesBehaviour
import org.pyjjs.scheduler.core.api.impl.actors.system.behaviours.TaskAppearedBehaviour

class SchedulingController(listener: SchedulingListener, registryOfStrategies: RegistryOfStrategies) : BehaviourBasedActor<SchedulingControllerState>() {

    init {
        val schedulingControllerState = SchedulingControllerState(context, registryOfStrategies)
        schedulingControllerState.schedulingListeners.add(listener)
        updateActorState(schedulingControllerState)
    }

    override fun init() {
        fillBehaviours()
        LOG.info("Scheduling controller is initialized")
    }

    private fun fillBehaviours() {
        addBehaviour(GotNewChangesBehaviour::class)
        addBehaviour(NotifyAboutChangesBehaviour::class)
        addBehaviour(TaskAppearedBehaviour::class)
    }
}
